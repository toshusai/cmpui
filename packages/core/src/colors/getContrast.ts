import { getLuminance } from "./getLuminance";
import { RGB } from "./types/RGB";

export function getContrast(rgb1: RGB, rgb2: RGB): number {
  const luminance1 = getLuminance(rgb1);
  const luminance2 = getLuminance(rgb2);
  const brightest = Math.max(luminance1, luminance2);
  const darkest = Math.min(luminance1, luminance2);
  return (brightest + 0.05) / (darkest + 0.05);
}
