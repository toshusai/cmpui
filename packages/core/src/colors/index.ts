export * from "./blendColors";
export * from "./chooseHigherContrast";
export * from "./componentToHex";
export * from "./const";
export * from "./getContrast";
export * from "./getHighContrast";
export * from "./getLuminance";

export * from "./hexToHsv";

export * from "./hexToRgba";
export * from "./hsvToHex";
export * from "./hsvToRgb";
export * from "./hsvaToRgba";
export * from "./isValidColor";
export * from "./isValidHex";
export * from "./isValidHex6";
export * from "./rgbToCss";
export * from "./rgbToHex";

export * from "./rgbToHsv";
export * from "./rgbaArrayToRgba";
export * from "./rgbaToCss";
export * from "./srgbToLinear";
export * from "./types";

export * from "./useHighContrastColor"