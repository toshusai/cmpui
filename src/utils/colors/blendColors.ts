import { RGB } from "./types/RGB";

export function blendColors(
  foregroundColor: RGB,
  backgroundColor: RGB,
  alpha: number,
): RGB {
  return {
    r: Math.round(foregroundColor.r * alpha + backgroundColor.r * (1 - alpha)),
    g: Math.round(foregroundColor.g * alpha + backgroundColor.g * (1 - alpha)),
    b: Math.round(foregroundColor.b * alpha + backgroundColor.b * (1 - alpha)),
  };
}
