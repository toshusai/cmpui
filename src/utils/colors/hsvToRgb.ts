import { HSV, RGB } from "./types";

/**
 * Reference:
 * - Algorithm from: https://en.wikipedia.org/wiki/HSL_and_HSV#HSV_to_RGB
 */
export function hsvToRgb(hsv: HSV): RGB {
  let r = 0;
  let g = 0;
  let b = 0;

  const h = hsv.h;
  const s = hsv.s;
  const v = hsv.v;

  const i = Math.floor(h / 60);
  const f = h / 60 - i;
  const p = v * (1 - s);
  const q = v * (1 - s * f);
  const t = v * (1 - s * (1 - f));

  switch (i % 6) {
    case 0:
      r = v;
      g = t;
      b = p;
      break;
    case 1:
      r = q;
      g = v;
      b = p;
      break;
    case 2:
      r = p;
      g = v;
      b = t;
      break;
    case 3:
      r = p;
      g = q;
      b = v;
      break;
    case 4:
      r = t;
      g = p;
      b = v;
      break;
    case 5:
      r = v;
      g = p;
      b = q;
      break;
  }

  return {
    r: Math.round(r * 255),
    g: Math.round(g * 255),
    b: Math.round(b * 255),
  };
}
