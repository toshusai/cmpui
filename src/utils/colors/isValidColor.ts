export function isValidColor(color: string) {
  const s = new Option().style;
  s.color = color;
  return s.color !== "";
}
