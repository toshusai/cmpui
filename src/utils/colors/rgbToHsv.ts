import { HSV, RGB } from "./types";

/**
 *
 * References:
 * - https://gist.github.com/mjackson/5311256
 * - https://en.wikipedia.org/wiki/HSL_and_HSV
 * @param r
 * @param g
 * @param b
 * @returns
 */
export function rgbToHsv(rgb: RGB): HSV {
  const red = rgb.r / 255;
  const green = rgb.g / 255;
  const blue = rgb.b / 255;

  const max = Math.max(red, green, blue);
  const min = Math.min(red, green, blue);
  const value = max;

  const chroma = max - min;
  const saturation = value === 0 ? 0 : chroma / value;

  let hue = 0;
  if (chroma === 0) {
    hue = 0;
  } else {
    switch (value) {
      case red:
        hue = (green - blue) / chroma + (green < blue ? 6 : 0);
        break;
      case green:
        hue = (blue - red) / chroma + 2;
        break;
      case blue:
        hue = (red - green) / chroma + 4;
        break;
    }
    hue /= 6;
  }

  return { h: hue * 360, s: saturation, v: value };
}
