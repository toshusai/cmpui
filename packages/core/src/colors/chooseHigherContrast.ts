import { getContrast } from "./getContrast";
import { RGB } from "./types/RGB";

export function chooseHigherContrast(
  baseColor: RGB,
  color1: RGB,
  color2: RGB,
  bias = 0,
): RGB {
  const contrast1 = getContrast(baseColor, color1);
  const contrast2 = getContrast(baseColor, color2);
  if (contrast1 > contrast2 + bias) {
    return color1;
  }
  return contrast1 > contrast2 ? color1 : color2;
}
