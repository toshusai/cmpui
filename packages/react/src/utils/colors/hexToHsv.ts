import { hexToRgba } from "./hexToRgba";
import { rgbToHsv } from "./rgbToHsv";
import { HSVA } from "./types/HSVA";

export function hexToHsv(hex: string): HSVA {
  const rgb = hexToRgba(hex);
  return { ...rgbToHsv(rgb), a: rgb.a };
}
