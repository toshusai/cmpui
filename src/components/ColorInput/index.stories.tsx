import { Meta, StoryObj } from "@storybook/react";
import { useState } from "react";
import { ColorInput } from ".";
import { hexToHsv } from "../../utils/colors/hexToHsv";

const meta: Meta<typeof ColorInput> = {
  title: "Components/Colors/ColorInput",
  component: ColorInput,
};

export default meta;
type Story = StoryObj<typeof ColorInput>;

export const Primary: Story = {
  render: function RenderBasic() {
    const [color, setColor] = useState({ ...hexToHsv("#c649ba"), a: 0.5 });

    return (
      <div
        style={{
          width: 192,
        }}
      >
        <ColorInput
          label="ColorInput"
          value={color}
          onChange={(value) => setColor(value)}
        />
      </div>
    );
  },
};
