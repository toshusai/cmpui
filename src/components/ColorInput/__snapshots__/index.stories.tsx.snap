// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Colors/ColorInput Primary smoke-test 1`] = `
<div style="width: 192px;">
  <div class="cmpui_color-input__root">
    <div class="cmpui_text-input__root">
      <div class="alpha-checker-board"
           style="--cmpui-block-size: 4px;"
      >
        <div tabindex="0"
             class="cmpui_color-input__color"
             style="background-color: rgb(198, 73, 186); mask-image: linear-gradient(to right, rgba(0, 0, 0, 0.5) 50%, black 50%);"
        >
        </div>
      </div>
      <label class="cmpui_text-input__label">
        ColorInput
      </label>
      <input type="text"
             class="cmpui_text-input__input"
             placeholder
             value="#c649ba"
      >
    </div>
    <div class="cmpui_text-input__root cmpui_slider-number-input__root">
      <label class="cmpui_text-input__label cmpui_slider-number-input__label">
        %
      </label>
      <input type="text"
             class="cmpui_text-input__input"
             placeholder
             value="50"
      >
    </div>
  </div>
</div>
`;
